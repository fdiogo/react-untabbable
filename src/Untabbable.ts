import { cloneElement, useRef } from 'react';

import useUntabbable from './useUntabbable';

const Untabbable: React.FunctionComponent<Props> = (props) => {
    const { children, disabled, includeContainer } = props;

    const ref = useRef<HTMLElement>(null);

    useUntabbable(ref, { disabled, includeContainer });

    return cloneElement(children, { ref });
}

export interface Props {
    /** 
     * Disable the component and restore the tabbable behaviour.
     * This is usually preferable to remove the component from tree
     * as it would trigger a re-render.
     */
    disabled?: boolean,
    /**
     * If it should also make the container that it wraps untabbable.
     */
    includeContainer?: boolean,
    children: React.ReactElement,
}

export default Untabbable;
