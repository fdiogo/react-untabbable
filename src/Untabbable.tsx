import React, { Children, createRef, cloneElement, isValidElement } from 'react';
import useUntabbable from './useUntabbable';

import type { FC, RefObject } from 'react';
import type { Options } from './types';

const Untabbable: FC<Pick<Options, 'disabled'>> = (props) => {
    const { children, disabled } = props;

    const childrenAsArray = Children.toArray(children);

    const refs: Record<number, RefObject<HTMLElement>> = {};
    childrenAsArray.forEach((child, index) => {
        if (isValidElement(child)) {
            refs[index] = createRef<HTMLElement>();
        }
    });

    useUntabbable(Object.values(refs), { disabled });

    return (
        <>
            {childrenAsArray.map((child, index) => {
                if (isValidElement(child)) {
                    return cloneElement(child, { ref: refs[index] });
                }
        
                return child;
            })}
        </>
    );
}

export default Untabbable;
