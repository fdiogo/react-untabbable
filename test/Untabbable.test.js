import React from 'react';
import { render } from '@testing-library/react';

import Untabbable from '../src/Untabbable';

// We mock tabbable here since what we are really interested in
// testing is the toggling of the tabindex.
jest.mock('tabbable', () => (element, options) => {
    const tabbableElements = [];

    if(options.includeContainer) {
        tabbableElements.push(element);
    }

    tabbableElements.push(...element.children);

    return tabbableElements;
});

describe('Untabbable', () => {
    it('should set the focusable children with tabindex=-1', () => {
        const { container, getByText } = render(
            <Untabbable>
                <div>
                    <button>First</button>
                    <button>Second</button>
                </div>
            </Untabbable>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        expect(container.firstChild.getAttribute("tabindex")).not.toBe("-1");
        expect(firstButton.getAttribute("tabindex")).toBe("-1");
        expect(secondButton.getAttribute("tabindex")).toBe("-1");
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should work with multiple children', () => {
        const { container, getByText } = render(
            <Untabbable>
                <div>
                    <button>First</button>
                </div>
                <div>
                    <button>Second</button>
                </div>
            </Untabbable>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        const firstDiv = container.children.item(0);
        const secondDiv = container.children.item(1);

        expect(firstDiv.getAttribute("tabindex")).not.toBe("-1");
        expect(firstButton.getAttribute("tabindex")).toBe("-1");

        expect(secondDiv.getAttribute("tabindex")).not.toBe("-1");
        expect(secondButton.getAttribute("tabindex")).toBe("-1");
    });

    it('should restore the previous tabindex when disabled', () => {
        const { container, getByText } = render(
            <Untabbable>
                <div>
                    <button tabIndex="2">First</button>
                    <button>Second</button>
                </div>
            </Untabbable>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        expect(firstButton.getAttribute("tabindex")).toBe("-1");
        expect(secondButton.getAttribute("tabindex")).toBe("-1");

        render(
            <Untabbable disabled>
                <div>
                    <button tabIndex="2">First</button>
                    <button>Second</button>
                </div>
            </Untabbable>,
        { container });

        expect(firstButton.getAttribute("tabindex")).toBe("2");
        expect(secondButton.getAttribute("tabindex")).toBe(null);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should also disable the container when `includeContainer` is true', () => {
        const { container, getByText } = render(
            <Untabbable includeContainer>
                <div>
                    <button>First</button>
                    <button>Second</button>
                </div>
            </Untabbable>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        expect(container.firstChild.getAttribute("tabindex")).toBe("-1");
        expect(firstButton.getAttribute("tabindex")).toBe("-1");
        expect(secondButton.getAttribute("tabindex")).toBe("-1");
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should treat all direct children as containers', () => {
        const { container, getByText } = render(
            <Untabbable includeContainer>
                <div>
                    <button>First</button>
                </div>
                <div>
                    <button>Second</button>
                </div>
            </Untabbable>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        const firstDiv = container.children.item(0);
        const secondDiv = container.children.item(1);

        expect(firstDiv.getAttribute("tabindex")).toBe("-1");
        expect(firstButton.getAttribute("tabindex")).toBe("-1");
        
        expect(secondButton.getAttribute("tabindex")).toBe("-1");
        expect(secondDiv.getAttribute("tabindex")).toBe("-1");

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should not only support elements but also any type of node', () => {
        const { container, getByText } = render(
            <Untabbable includeContainer>
                <div>
                    <button>First</button>
                </div>
                Hello!
                <div>
                    <button>Second</button>
                </div>
                {1}
                {false}
                {null}
            </Untabbable>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        const firstDiv = container.children.item(0);
        const secondDiv = container.children.item(1);

        expect(firstDiv.getAttribute("tabindex")).toBe("-1");
        expect(firstButton.getAttribute("tabindex")).toBe("-1");
        
        expect(secondButton.getAttribute("tabindex")).toBe("-1");
        expect(secondDiv.getAttribute("tabindex")).toBe("-1");

        expect(container.firstChild).toMatchSnapshot();
    });

    it('should not alter the tabindex when `disabled` is true', () => {
        const { container, getByText } = render(
            <Untabbable disabled>
                <div>
                    <button>First</button>
                    <button>Second</button>
                </div>
            </Untabbable>
        );

        const firstButton = getByText('First');
        const secondButton = getByText('Second');

        expect(container.firstChild.getAttribute("tabindex")).not.toBe("-1");
        expect(firstButton.getAttribute("tabindex")).not.toBe("-1");
        expect(secondButton.getAttribute("tabindex")).not.toBe("-1");
        expect(container.firstChild).toMatchSnapshot();
    });
});