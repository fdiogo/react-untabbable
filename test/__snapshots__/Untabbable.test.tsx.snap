// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Untabbable should also disable the container when \`includeContainer\` is true 1`] = `
<div
  tabindex="-1"
>
  <button
    tabindex="-1"
  >
    First
  </button>
  <button
    tabindex="-1"
  >
    Second
  </button>
</div>
`;

exports[`Untabbable should not alter the tabindex when \`disabled\` is true 1`] = `
<div
  tabindex="0"
>
  <button>
    First
  </button>
  <button>
    Second
  </button>
</div>
`;

exports[`Untabbable should not only support elements but also any type of node 1`] = `
<div
  tabindex="-1"
>
  <button
    tabindex="-1"
  >
    First
  </button>
</div>
`;

exports[`Untabbable should restore the previous tabindex when disabled 1`] = `
<div
  tabindex="0"
>
  <button
    tabindex="2"
  >
    First
  </button>
  <button>
    Second
  </button>
</div>
`;

exports[`Untabbable should set the focusable children with tabindex=-1 1`] = `
<div
  tabindex="-1"
>
  <button
    tabindex="-1"
  >
    First
  </button>
  <button
    tabindex="-1"
  >
    Second
  </button>
</div>
`;

exports[`Untabbable should treat all direct children as containers 1`] = `
<div
  tabindex="-1"
>
  <button
    tabindex="-1"
  >
    First
  </button>
</div>
`;
